### PR TITLE
Restart the projection subsystem

### PR DESCRIPTION
**Pull request by hayley-jean**
Add an endpoint to the projections controller that allows restarting the projection subsystem.

Handle the restart message in the projection subsystem, which publishes a start and stop component messages.
Handle master and system ready changes in the `ProjectionsSubsystem` rather than in both `ProjectionManager` and `ProjectionCoreCoordinator`.

Use a generated Correlation Id instead of the epoch id to differentiate between each run of projections.
The Correlation Id is included in the StartComponent message published by the ProjectionsSubsystem.

Add a new core projection state, Suspended, that unsubscribes the projection's event readers when entered.

Track the state of `ProjectionCoreCoordinator` and `ProjectionManager` using an enum of possible states rather than a few bools.

Try and coordinate the projection shutdown a bit better:

**ProjectionCoreCoordinator:**

1. Publish the StopCore message to start suspending the projections
2. Track which projections should be suspended when suspending them, and listen for the projection's "Suspended" message before considering the ProjectionCoreSystem as fully stopped.
3. When ProjectionCoreService is stopped, stop the readers for that queue.
4. The subscriptions and event readers should have been cleaned up in the EventReaderCoreService with the projections being suspended in step 2.

If the projections take more than 5 seconds to enter the `suspended` state in the ProjectionCoreService, it will force stop the service.

**ProjectionManager:**
1. Stops itself
2. Clears the projections
3. Starts again.

**Note**
This affects the subsystem starting and stopping on master state change.
Testing should include resigning master and master changeovers in a cluster.

### CODE REVIEW

**Comment by hayley-jean** [comment]
### Issues made more pronounced by this PR:

#### 1. Projection Manager Response Reader

The `ProjectionManagerResponseReader` has no stop functionality.
If the projections subsystem starts on a node, the response reader will always be running.
It will keep reading the projection manager stream and publishing messages on the projection bus.
The other components aren't running, so these messages aren't acted upon.
This seems harmless at the moment, but should be fixed.

To see this behaviour:
1. Start a cluster with projections running.
2. Resign the master node such that a new node takes over.
3. Observe the `$prepared`, `$starting`, and `$started` message logs on both the old and new msater nodes.

#### 2. Emitting projections look like they're stuck

Possibly related to https://github.com/EventStore/EventStore/issues/2001

It's possible for the projections to get into a state where they look like they're stuck, but will recover given enough time.

This is caused by the projection processing more events than it's checkpoint count quickly - for example, if it processes all of the events before the `CheckpointAfterMs` time has elapsed.
On a restart, the projection will need to catch up and verify these emitted events before it can start checkpointing again.

File: src/EventStore.Projections.Core/Services/Processing/RequestResponseQueueForwarder.cs [hunk 1]
```diff
@@ -13,7 +13,7 @@ public class RequestResponseQueueForwarder : IHandle<ClientMessage.ReadEvent>,
 		IHandle<SystemMessage.SubSystemInitialized>,
 		IHandle<ProjectionCoreServiceMessage.SubComponentStarted>,
 		IHandle<ProjectionCoreServiceMessage.SubComponentStopped> {
-		private readonly IPublisher _externalRequestQueue;
+	private readonly IPublisher _externalRequestQueue;
```
Comment (pgermishuys) [inline src/EventStore.Projections.Core/Services/Processing/RequestResponseQueueForwarder.cs]: Looks like a formatting issue.

File: src/EventStore.Projections.Core/Services/Management/ProjectionCoreCoordinator.cs [hunk 2]
```diff
@@ -11,22 +11,25 @@
...
-		private int _pendingSubComponentsStarts = 0;
-		private int _activeSubComponents = 0;
-		private bool _newInstanceWaiting = false;
+		private int _pendingSubComponentsStarts;
+		private int _activeSubComponents;
+		
+		private Guid _runCorrelationId = Guid.Empty;
+		private CoreCoordinatorState _currentState = CoreCoordinatorState.Stopped;
+		
+		private const string ComponentName = "ProjectionCoreCoordinator";
```
[the remark this answers is not in the document]
Comment (hayley-jean) [reply src/EventStore.Projections.Core/Services/Management/ProjectionCoreCoordinator.cs]: I agree. Do you prefer using a constant in the class itself, or using nameof for identifying the? e.g. `nameof(ProjectionCoreService)`

File: src/EventStore.Projections.Core.Tests/Services/command_reader_response_reader_integration/when_the_subsystem_components_start.cs [hunk 3]
```diff
@@ -1,37 +1,30 @@
...
-			yield return new SystemMessage.EpochWritten(epochRecord);
-			yield return new SystemMessage.SystemCoreReady();
+			yield return new ProjectionSubsystemMessage.StartComponents(_uniqueId);
 		}
 
 		[Test]
 		public void readers_should_use_the_same_control_stream_id() {
 			Assert.AreEqual(_uniqueId,
-				_consumer.HandledMessages.OfType<ProjectionManagementMessage.Starting>().First().EpochId);
+				_consumer.HandledMessages.OfType<ProjectionManagementMessage.Starting>().First().RunCorrelationId);
```
[the remark this answers is not in the document]
Comment (hayley-jean) [reply src/EventStore.Projections.Core.Tests/Services/command_reader_response_reader_integration/when_the_subsystem_components_start.cs]: :+1: `InstanceCorrelationId` is more descriptive of what it is

File: src/EventStore.Projections.Core.Tests/Services/core_coordinator/when_restarting_with_projection_type_all.cs [hunk 4]
```diff
@@ -1,12 +1,7 @@
 using System;
```
[the remark this answers is not in the document]
Comment (hayley-jean) [reply src/EventStore.Projections.Core.Tests/Services/core_coordinator/when_restarting_with_projection_type_all.cs]: Yup, that would make them clearer, I'll remove these methods

File: src/EventStore.Projections.Core.Tests/Services/core_coordinator/when_restarting_with_projection_type_all.cs [hunk 5]
```diff
@@ -28,122 +24,113 @@ public class when_restarting_with_projection_type_all {
...
-
-			_coordinator.Handle(
-				new ProjectionCoreServiceMessage.SubComponentStopped("ProjectionCoreServiceCommandReader"));
-
-			Assert.AreEqual(1, queues[0].Messages.FindAll(x => x is ReaderCoreServiceMessage.StartReader).Count);
-			Assert.AreEqual(1, queues[0].Messages.FindAll(x => x is ProjectionCoreServiceMessage.StartCore).Count);
+		public void should_not_stop_if_not_started() {
+			AllSubComponentsStopped(initialCorrelationId);
+			
+			StopComponents(Guid.NewGuid());
```
[the remark this answers is not in the document]
Comment (hayley-jean) [reply src/EventStore.Projections.Core.Tests/Services/core_coordinator/when_restarting_with_projection_type_all.cs]: Good catch, thanks!

File: src/EventStore.Projections.Core/Services/Management/ProjectionCoreCoordinator.cs [hunk 6]
```diff
@@ -114,61 +87,111 @@ public class ProjectionCoreCoordinator
...
+			}
 			_activeSubComponents--;
 			Log.Debug("PROJECTIONS: SubComponent Stopped: {subComponent}", message.SubComponent);
 
-			if (_newInstanceWaiting)
-				StartWhenConditionsAreMet();
+			if (message.SubComponent == "ProjectionCoreService") {
+				if (!_queues.TryGetValue(message.CorrelationId, out var queue))
+					return;
+				queue.Publish(new ReaderCoreServiceMessage.StopReader(message.CorrelationId));
```
[the remark this answers is not in the document]
Comment (hayley-jean) [reply src/EventStore.Projections.Core/Services/Management/ProjectionCoreCoordinator.cs]: The Id in `StopReader` isn't used directly, but it's required by the `SubComponentStopped` message that gets published when `StopReader` is handled

**Comment by hayley-jean** [comment]
Fixes #2014

**Pull request merged**